src/mongoChangeStream.ts

```diff
@@ -463,37 +475,57 @@ export function initSync<ExtendedEvents extends EventEmitter.ValidEventTypes>(
         timeout: ms('30s'),
         ...options,
       })
-      // Start the change stream
-      changeStream = await getChangeStream()
+
       const nextChecker = safelyCheckNext(changeStream)
       state.change('started')
-      let event: ChangeStreamDocument | null
-      // Consume change stream
-      while (await nextChecker.hasNext()) {
-        event = await changeStream.next()
-        debug('Change stream event %O', event)
-        // Omit nested fields that are not handled by $unset.
-        // For example, if 'a' was omitted then 'a.b.c' should be omitted.
-        if (
-          omit &&
-          event.operationType === 'update' &&
-          // Downstream libraries might unset event.updateDescription
-          // to optimize performance (e.g., mongo2elastic).
-          event.updateDescription
-        ) {
-          omitFieldsForUpdate(omit, event)
+
+      // Consume change stream until there is an error or `stop` is called.
+      while (!state.is('stopping')) {
+        // This always updates `changeStream.resumeToken`, even when we reach
+        // the end of the change stream and null is returned.
+        //
+        // This is important because we always want to keep a recent resume
+        // token. Attempting to resume from an older resume token might not
+        // work, because only a certain (configurable) amount of oplog history
+        // is kept. Even if the oplog entry still exists, it can take a very
+        // long time to scan through all of the oplog entries since then, so
+        // keeping a recent resume token is vital for performance.
+        const event: ChangeStreamDocument | null = await nextChecker.getNext()
+
+        if (event) {
+          debug('Change stream event %O', event)
+          // Omit nested fields that are not handled by $unset.
+          // For example, if 'a' was omitted then 'a.b.c' should be omitted.
+          if (
+            omit &&
+            event.operationType === 'update' &&
+            // Downstream libraries might unset event.updateDescription
+            // to optimize performance (e.g., mongo2elastic).
+            event.updateDescription
+          ) {
+            omitFieldsForUpdate(omit, event)
+          }
         }
+
         await queue.enqueue(event)
         await pause.maybeBlock()
+
+        if (!event) {
+          // The null event (i.e. "end of change stream") is always the last
+          // event in the batch. See `_processRecords`.
+          await queue.flush()
+        }
```

Comment:
Can likely remove this. Should adjust `_processRecords`, since the null record won't necessarily be the last one.

Reply:
Batch queue timeout should ensure that we process the null entries in a reasonable amount of time.